File: packages/kl-crewai-tools/src/kl_crewai_tools/tools.py
"""CrewAI ``BaseTool`` subclasses for kanban-lite operations.

Each tool wraps a single REST API call via :class:`KanbanLiteClient`.
Input validation is handled by Pydantic models (``args_schema``).

Usage::

    from kl_crewai_tools import KanbanLiteClient, ListCardsTool

    client = KanbanLiteClient("http://localhost:3000")
    tool = ListCardsTool(client=client)
    result = tool.run()
"""

from __future__ import annotations

import json
from typing import Any, Optional, Type

from crewai.tools import BaseTool
from pydantic import BaseModel, Field

from kl_crewai_tools.client import KanbanLiteClient


# ------------------------------------------------------------------ #
# Input schemas                                                        #
# ------------------------------------------------------------------ #


class ListCardsInput(BaseModel):
    """Input schema for listing kanban cards."""

    status: Optional[str] = Field(
        default=None,
        description="Filter by status column, e.g. backlog | todo | in-progress | review | done",
    )
    priority: Optional[str] = Field(
        default=None,
        description="Filter by priority level: critical | high | medium | low",
    )
    assignee: Optional[str] = Field(
        default=None, description="Filter by assignee name"
    )
    label: Optional[str] = Field(
        default=None, description="Filter by label name"
    )


class GetCardInput(BaseModel):
    """Input schema for getting a single card."""

    card_id: str = Field(description="Card ID or partial card ID")


class CreateCardInput(BaseModel):
    """Input schema for creating a card."""

    content: str = Field(
        description=(
            "Markdown content for the card. "
            "Must start with a # heading that becomes the card title, "
            "e.g. '# Fix login regression\\n\\nDetails here.'"
        )
    )
    status: Optional[str] = Field(
        default=None,
        description="Initial status column (defaults to board default)",
    )
    priority: Optional[str] = Field(
        default=None,
        description="Priority level: critical | high | medium | low",
    )
    assignee: Optional[str] = Field(
        default=None, description="Assigned team member"
    )
    labels: Optional[str] = Field(
        default=None,
        description="Comma-separated label names, e.g. 'bug,frontend'",
    )
    due_date: Optional[str] = Field(
        default=None, description="Due date in ISO 8601 format, e.g. 2026-04-01"
    )


class UpdateCardInput(BaseModel):
    """Input schema for updating a card."""

    card_id: str = Field(description="Card ID or partial card ID")
    content: Optional[str] = Field(
        default=None, description="Updated Markdown content"
    )
    priority: Optional[str] = Field(
        default=None,
        description="Priority level: critical | high | medium | low",
    )
    assignee: Optional[str] = Field(
        default=None, description="Assigned team member"
    )
    labels: Optional[str] = Field(
        default=None,
        description="Comma-separated label names, e.g. 'bug,frontend'",
    )
    due_date: Optional[str] = Field(
        default=None, description="Due date in ISO 8601 format"
    )


class MoveCardInput(BaseModel):
    """Input schema for moving a card."""

    card_id: str = Field(description="Card ID or partial card ID")
    target_column: str = Field(
        description="Target status column, e.g. todo | in-progress | review | done"
    )
    position: Optional[int] = Field(
        default=None,
        description="Zero-based position within the target column (default: top)",
    )


class DeleteCardInput(BaseModel):
    """Input schema for deleting a card."""

    card_id: str = Field(description="Card ID or partial card ID")


class ListColumnsInput(BaseModel):
    """Input schema for listing columns (no parameters needed)."""


class GetCommentsInput(BaseModel):
    """Input schema for getting comments on a card."""

    card_id: str = Field(description="Card ID or partial card ID")


class AddCommentInput(BaseModel):
    """Input schema for adding a comment to a card."""

    card_id: str = Field(description="Card ID or partial card ID")
    content: str = Field(description="Markdown comment body")
    author: Optional[str] = Field(
        default=None,
        description="Display name for the comment author (default: crewai-agent)",
    )


# ------------------------------------------------------------------ #
# Tool helpers                                                         #
# ------------------------------------------------------------------ #


def _format(data: Any) -> str:
    """Return a JSON string from API response data."""
    return json.dumps(data, indent=2, default=str)


def _parse_labels(raw: str | None) -> list[str] | None:
    """Split comma-separated label string into a list."""
    if not raw:
        return None
    return [l.strip() for l in raw.split(",") if l.strip()]


# ------------------------------------------------------------------ #
# CrewAI Tools                                                         #
# ------------------------------------------------------------------ #


class ListCardsTool(BaseTool):
    """List kanban cards on the default board with optional filters."""

    name: str = "list_cards"
    description: str = (
        "List kanban cards on the default board. "
        "Optionally filter by status column, priority, assignee, or label."
    )
    args_schema: Type[BaseModel] = ListCardsInput
    client: KanbanLiteClient

    def _run(
        self,
        status: str | None = None,
        priority: str | None = None,
        assignee: str | None = None,
        label: str | None = None,
    ) -> str:
        cards = self.client.list_cards(
            status=status, priority=priority, assignee=assignee, label=label
        )
        return _format(cards)


class GetCardTool(BaseTool):
    """Retrieve full details of a single kanban card by its ID."""

    name: str = "get_card"
    description: str = (
        "Retrieve full details of a single kanban card by its ID or partial ID."
    )
    args_schema: Type[BaseModel] = GetCardInput
    client: KanbanLiteClient

    def _run(self, card_id: str) -> str:
        card = self.client.get_card(card_id)
        return _format(card)


class CreateCardTool(BaseTool):
    """Create a new kanban card on the default board."""

    name: str = "create_card"
    description: str = (
        "Create a new kanban card on the default board. "
        "Content must start with a # heading that becomes the card title."
    )
    args_schema: Type[BaseModel] = CreateCardInput
    client: KanbanLiteClient

    def _run(
        self,
        content: str,
        status: str | None = None,
        priority: str | None = None,
        assignee: str | None = None,
        labels: str | None = None,
        due_date: str | None = None,
    ) -> str:
        card = self.client.create_card(
            content=content,
            status=status,
            priority=priority,
            assignee=assignee,
            labels=_parse_labels(labels),
            due_date=due_date,
        )
        return _format(card)


class UpdateCardTool(BaseTool):
    """Update fields on an existing kanban card."""

    name: str = "update_card"
    description: str = (
        "Update fields on an existing kanban card. "
        "Only supplied fields are changed; omitted fields are left unchanged."
    )
    args_schema: Type[BaseModel] = UpdateCardInput
    client: KanbanLiteClient

    def _run(
        self,
        card_id: str,
        content: str | None = None,
        priority: str | None = None,
        assignee: str | None = None,
        labels: str | None = None,
        due_date: str | None = None,
    ) -> str:
        fields: dict[str, Any] = {}
        if content is not None:
            fields["content"] = content
        if priority is not None:
            fields["priority"] = priority
        if assignee is not None:
            fields["assignee"] = assignee
        if labels is not None:
            fields["labels"] = _parse_labels(labels)
        if due_date is not None:
            fields["dueDate"] = due_date
        card = self.client.update_card(card_id, **fields)
        return _format(card)


class MoveCardTool(BaseTool):
    """Move a kanban card to a different status column."""

    name: str = "move_card"
    description: str = "Move a kanban card to a different status column."
    args_schema: Type[BaseModel] = MoveCardInput
    client: KanbanLiteClient

    def _run(
        self,
        card_id: str,
        target_column: str,
        position: int | None = None,
    ) -> str:
        result = self.client.move_card(card_id, target_column, position=position)
        return f"Card {card_id} moved to {target_column}"


class DeleteCardTool(BaseTool):
    """Soft-delete a kanban card."""

    name: str = "delete_card"
    description: str = "Soft-delete a kanban card by its ID."
    args_schema: Type[BaseModel] = DeleteCardInput
    client: KanbanLiteClient

    def _run(self, card_id: str) -> str:
        self.client.delete_card(card_id)
        return f"Card {card_id} deleted"


class ListColumnsTool(BaseTool):
    """List all columns on the default board."""

    name: str = "list_columns"
    description: str = "List all columns (status lanes) on the default kanban board."
    args_schema: Type[BaseModel] = ListColumnsInput
    client: KanbanLiteClient

    def _run(self) -> str:
        columns = self.client.list_columns()
        return _format(columns)


class GetCommentsTool(BaseTool):
    """Retrieve all comments for a kanban card."""

    name: str = "get_comments"
    description: str = "Retrieve all comments for a kanban card by its ID."
    args_schema: Type[BaseModel] = GetCommentsInput
    client: KanbanLiteClient

    def _run(self, card_id: str) -> str:
        comments = self.client.get_comments(card_id)
        return _format(comments)


class AddCommentTool(BaseTool):
    """Add a comment to a kanban card."""

    name: str = "add_comment"
    description: str = "Add a comment to a kanban card."
    args_schema: Type[BaseModel] = AddCommentInput
    client: KanbanLiteClient

    def _run(
        self,
        card_id: str,
        content: str,
        author: str | None = None,
    ) -> str:
        comment = self.client.add_comment(
            card_id, content, author=author or "crewai-agent"
        )
        return _format(comment)

File: packages/kl-adapter-vercel-ai/src/tools.ts
// ---------------------------------------------------------------------------
// Vercel AI SDK tool definitions for kanban-lite.
//
// Exports a factory function that accepts a KanbanClient (or config) and
// returns a record of `tool()` definitions ready for use with `streamText()`
// or `generateText()` from the Vercel AI SDK.
// ---------------------------------------------------------------------------

import { tool } from 'ai'
import { z } from 'zod'
import { KanbanClient } from './client'
import type { KanbanClientConfig } from './types'

// ---------------------------------------------------------------------------
// Serialisation helpers (same as in the example route.ts)
// ---------------------------------------------------------------------------

function serializeActions(actions: string[] | Record<string, string> | undefined) {
  if (!actions) return []
  if (Array.isArray(actions)) {
    return actions.map((key) => ({ key, title: key }))
  }
  return Object.entries(actions).map(([key, title]) => ({ key, title }))
}

function serializeForms(forms: { name?: string; schema?: Record<string, unknown> }[] | undefined) {
  return (forms ?? []).map((form, index) => ({
    id: form.name ?? String(form.schema?.title ?? `form-${index + 1}`),
    name: form.name ?? String(form.schema?.title ?? `Form ${index + 1}`),
  }))
}

// ---------------------------------------------------------------------------
// Tool set configuration
// ---------------------------------------------------------------------------

/** Options for customising the generated tool set. */
export interface KanbanToolsOptions {
  /**
   * Maximum number of cards returned by the `list_cards` tool.
   * @default 50
   */
  listLimit?: number
  /**
   * Maximum number of recent comments included in the `get_card` tool result.
   * @default 20
   */
  commentLimit?: number
  /**
   * Default author name used when the LLM does not provide one.
   * @default 'kanban-chat-agent'
   */
  defaultAuthor?: string
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/**
 * Create a complete set of Vercel AI SDK tool definitions for kanban-lite.
 *
 * The returned object can be spread directly into the `tools` option of
 * `streamText()` or `generateText()`.
 *
 * @example
 * ```ts
 * import { streamText } from 'ai'
 * import { openai } from '@ai-sdk/openai'
 * import { createKanbanTools } from 'kl-adapter-vercel-ai'
 *
 * const tools = createKanbanTools({
 *   baseUrl: 'http://localhost:3000',
 *   boardId: 'default',
 * })
 *
 * const result = streamText({
 *   model: openai('gpt-4o-mini'),
 *   tools,
 *   messages,
 * })
 * ```
 */
export function createKanbanTools(
  clientOrConfig: KanbanClient | KanbanClientConfig = {},
  options: KanbanToolsOptions = {},
) {
  const client =
    clientOrConfig instanceof KanbanClient
      ? clientOrConfig
      : new KanbanClient(clientOrConfig)

  const listLimit = options.listLimit ?? 50
  const commentLimit = options.commentLimit ?? 20
  const defaultAuthor = options.defaultAuthor ?? 'kanban-chat-agent'

  return {
    // -------------------------------------------------------------------
    // Cards
    // -------------------------------------------------------------------

    create_card: tool({
      description:
        'Create a new kanban card. Use when the user asks to add, log, or create a task.',
      parameters: z.object({
        title: z.string().describe('Short, clear card title'),
        description: z
          .string()
          .optional()
          .describe('Optional detailed description of the card'),
        assignee: z.string().optional().describe('Optional assignee name'),
        priority: z
          .enum(['critical', 'high', 'medium', 'low'])
          .default('medium')
          .describe('Card priority level'),
        status: z
          .string()
          .optional()
          .describe('Initial column, e.g. "backlog", "in-progress". Defaults to the board default.'),
        labels: z
          .array(z.string())
          .optional()
          .describe('Optional labels to assign to the card'),
        actionKeys: z
          .array(z.string())
          .optional()
          .describe('Optional card action keys such as deploy, rollback, notify-slack'),
        formTemplates: z
          .array(z.string())
          .optional()
          .describe('Optional reusable form ids to attach to the card'),
      }),
      execute: async ({ title, description, assignee, priority, status, labels, actionKeys, formTemplates }) => {
        try {
          const card = await client.createCard(title, description, priority, {
            ...(assignee ? { assignee } : {}),
            ...(status ? { status } : {}),
            ...(labels?.length ? { labels } : {}),
            ...(actionKeys?.length ? { actions: actionKeys } : {}),
            ...(formTemplates?.length
              ? { forms: formTemplates.map((name) => ({ name })) }
              : {}),
          })
          return {
            ok: true as const,
            card: {
              id: card.id,
              title: card.title,
              status: card.status,
              priority: card.priority,
              labels: card.labels ?? [],
              actions: serializeActions(card.actions),
              forms: serializeForms(card.forms),
            },
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    list_cards: tool({
      description:
        'List cards from the kanban board. Optionally filter by status column.',
      parameters: z.object({
        status: z
          .string()
          .optional()
          .describe(
            'Column to filter by, e.g. "backlog", "in-progress", "done". Omit to list all cards.',
          ),
      }),
      execute: async ({ status }) => {
        try {
          const cards = await client.listCards(status)
          return {
            ok: true as const,
            count: cards.length,
            cards: cards.slice(0, listLimit).map((card) => ({
              id: card.id,
              title: card.title,
              status: card.status,
              priority: card.priority,
              assignee: card.assignee ?? null,
              labels: card.labels ?? [],
              actions: serializeActions(card.actions),
              forms: serializeForms(card.forms),
              commentCount: card.comments?.length ?? 0,
            })),
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    get_card: tool({
      description:
        'Inspect a specific card to see its details, comments, attached forms, actions, and labels.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID to inspect'),
      }),
      execute: async ({ cardId }) => {
        try {
          const card = await client.getCard(cardId)
          return {
            ok: true as const,
            card: {
              id: card.id,
              title: card.title,
              status: card.status,
              priority: card.priority,
              assignee: card.assignee ?? null,
              dueDate: card.dueDate ?? null,
              labels: card.labels ?? [],
              metadata: card.metadata ?? {},
              actions: serializeActions(card.actions),
              forms: serializeForms(card.forms),
              formData: card.formData ?? {},
              comments: (card.comments ?? []).slice(-commentLimit),
              body: card.body ?? '',
            },
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    update_card: tool({
      description:
        'Update an existing card. Only provided fields are changed.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
        priority: z
          .enum(['critical', 'high', 'medium', 'low'])
          .optional()
          .describe('New priority level'),
        assignee: z.string().optional().describe('New assignee name'),
        labels: z
          .array(z.string())
          .optional()
          .describe('Replace card labels'),
        dueDate: z.string().optional().describe('New due date (ISO 8601)'),
      }),
      execute: async ({ cardId, priority, assignee, labels, dueDate }) => {
        try {
          const updates: Record<string, unknown> = {}
          if (priority !== undefined) updates.priority = priority
          if (assignee !== undefined) updates.assignee = assignee
          if (labels !== undefined) updates.labels = labels
          if (dueDate !== undefined) updates.dueDate = dueDate
          const card = await client.updateCard(cardId, updates)
          return {
            ok: true as const,
            card: {
              id: card.id,
              title: card.title,
              status: card.status,
              priority: card.priority,
              assignee: card.assignee ?? null,
              labels: card.labels ?? [],
            },
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    move_card: tool({
      description:
        'Move an existing card to a different status column. Partial card ID is supported.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
        status: z
          .string()
          .describe('Target column name, e.g. "in-progress", "review", "done"'),
      }),
      execute: async ({ cardId, status }) => {
        try {
          const card = await client.moveCard(cardId, status)
          return {
            ok: true as const,
            card: { id: card.id, title: card.title, status: card.status },
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    delete_card: tool({
      description: 'Delete (soft-delete) a card from the board.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID to delete'),
      }),
      execute: async ({ cardId }) => {
        try {
          await client.deleteCard(cardId)
          return { ok: true as const, cardId }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    // -------------------------------------------------------------------
    // Comments
    // -------------------------------------------------------------------

    add_comment: tool({
      description:
        'Add a markdown comment to a card. Use for freeform notes, updates, or requests tied to a specific card.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
        content: z.string().describe('Markdown comment body'),
        author: z
          .string()
          .optional()
          .default(defaultAuthor)
          .describe('Display name for the comment author'),
      }),
      execute: async ({ cardId, content, author }) => {
        try {
          const comment = await client.addComment(cardId, author, content)
          return { ok: true as const, comment }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    stream_comment: tool({
      description:
        'Stream a comment to a card so connected viewers see it arrive incrementally. Use this instead of add_comment when the text is long or the comment should appear in real-time.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
        content: z.string().describe('Full comment text to stream (supports markdown)'),
        author: z
          .string()
          .optional()
          .default(defaultAuthor)
          .describe('Display name for the comment author'),
      }),
      execute: async ({ cardId, content, author }) => {
        try {
          const comment = await client.streamComment(cardId, author, content)
          return { ok: true as const, comment }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    list_comments: tool({
      description: 'List all comments on a card.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
      }),
      execute: async ({ cardId }) => {
        try {
          const comments = await client.listComments(cardId)
          return { ok: true as const, count: comments.length, comments }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    // -------------------------------------------------------------------
    // Forms
    // -------------------------------------------------------------------

    submit_card_form: tool({
      description:
        'Submit structured data to a form attached to a card. Include form field values in the data object.',
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
        formId: z
          .string()
          .describe('Attached form id, e.g. incident-report or release-checklist'),
        data: z
          .record(z.string(), z.any())
          .describe('Submitted form field values as an object'),
      }),
      execute: async ({ cardId, formId, data }) => {
        try {
          const result = await client.submitCardForm(cardId, formId, data)
          return {
            ok: true as const,
            boardId: result.boardId,
            form: { id: result.form.id, label: result.form.label },
            data: result.data,
            card: {
              id: result.card.id,
              title: result.card.title,
              status: result.card.status,
            },
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    // -------------------------------------------------------------------
    // Actions
    // -------------------------------------------------------------------

    trigger_card_action: tool({
      description:
        "Trigger one of a card's named actions (e.g. deploy, rollback, notify-slack).",
      parameters: z.object({
        cardId: z.string().describe('Card ID or partial ID'),
        action: z.string().describe('Action key to trigger'),
      }),
      execute: async ({ cardId, action }) => {
        try {
          await client.triggerCardAction(cardId, action)
          const card = await client.getCard(cardId)
          return {
            ok: true as const,
            action,
            card: {
              id: card.id,
              title: card.title,
              status: card.status,
              priority: card.priority,
              commentCount: card.comments?.length ?? 0,
              actions: serializeActions(card.actions),
              forms: serializeForms(card.forms),
            },
          }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    // -------------------------------------------------------------------
    // Board info
    // -------------------------------------------------------------------

    get_board: tool({
      description:
        'Get board configuration including columns, actions, and metadata fields.',
      parameters: z.object({
        boardId: z
          .string()
          .optional()
          .describe('Board ID. Defaults to the configured board.'),
      }),
      execute: async ({ boardId }) => {
        try {
          const board = await client.getBoard(boardId)
          return { ok: true as const, board }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),

    list_columns: tool({
      description:
        'List the status columns configured on the board.',
      parameters: z.object({
        boardId: z
          .string()
          .optional()
          .describe('Board ID. Defaults to the configured board.'),
      }),
      execute: async ({ boardId }) => {
        try {
          const columns = await client.listColumns(boardId)
          return { ok: true as const, columns }
        } catch (err) {
          return { ok: false as const, error: String(err) }
        }
      },
    }),
  }
}
